### PR TITLE
Added capability to integrate delta funtions

### PR DESCRIPTION
**Pull request by stefanozampini**
LHS Integration of delta functions just accounts for sampling the basis at the center of the delta function.
This branch adds this capability to the DomainLFIntegrators

### CODE REVIEW

**Comment by stefanozampini** [comment]
@tzanio @v-dobrev Any reason why you have not yet approved this?

**Comment by v-dobrev** [comment]
I just have not had time to review it. I'll try to do it this week.

**Comment by v-dobrev** [comment]
@stefanozampini,

I think we can simplify the implementation by directly adding the `DeltaCoefficient`s to the `LinearForm` using two new methods (similar to the "Add*Integrator()" methods) - one for adding `DeltaCoefficient` and one for adding `VectorDeltaCoefficient`. This way there will be no need to modify the `LinearFormIntegrator` classes and no need to introduce new virtual methods in `Coefficient` and `VectorCoefficient`.

Basically, the `LinearForm` will have two additional arrays for `DeltaCoefficient`s and for `VectorDeltaCoefficient`s which will be treated separately in the `Assemble()` method the way the other arrays, `dlfi`, `blfi`, etc., are treated separately.

What do you think?

**Comment by stefanozampini** [comment]
What you are proposing was the alternative implementation I thought about before implementing this one. They both have drawbacks, and these come from the fact that the coefficients are two levels down than the linear forms, i.e. you create a LinearForm with a DomainLFIntegrator, and a DomainLFIntegrator with a Coefficient. I decided to go with the current one since I'd rather prefer to keep the same `LinearForm->AddDomainIntegrator(DomainLFIntegrator*)` interface, than having two new DomainLF*DeltaIntegrator classes and separate user code paths for *DeltaCoefficients.  I personally don't like specialized API unless it is actually needed.  What are the virtual methods you don't like?

**Comment by v-dobrev** [comment]
I don't like the new virtual methods in class `Coefficient`:
```c++
   virtual void GetDeltaCenter(Vector& center);
   virtual void AllowDeltaEval(const bool allow = true); 
```
and the new non-virtual and virtual methods in class `LinearFormIntegrator`:
```c++
   void CheckCoefficientIsDelta(Coefficient& Q);
   void CheckVectorCoefficientIsDelta(VectorCoefficient& Q);
   virtual bool IsDelta();
   virtual void GetDeltaCenter(Vector& center);
   virtual void AllowDeltaEval(const bool allow = true);
```
All of these are dealing with delta coefficients specifically but are defined at the highest abstraction level.

Let me think about it, maybe we can support delta coefficients with some `dynamic_cast`s in `AddDomainIntegrator`.

**Comment by stefanozampini** [comment]
I agree with you that these methods are quite horrible :-), but I could not come up with a better solution.

File: fem/linearform.cpp [hunk 1]
```diff
@@ -126,6 +154,69 @@ void LinearForm::Update(FiniteElementSpace *f, Vector &v, int v_offset)
...
+      fir[0].Set(ip.x,ip.y,ip.z,1.0);
+      dlfi[lfid]->SetIntRule(&fir);
+
+      // define fake ElementTransformation
+      FakeTransformation feltrans(fes->GetFE(elem_id)->GetDim());
+
+      // pointwise evaluation
+      fes -> GetElementVDofs (elem_id, vdofs);
+      dlfi[lfid]->AllowDeltaEval();
+      dlfi[lfid]->AssembleRHSElementVect(*fes->GetFE(elem_id), feltrans, elemvect);
```
Comment (v-dobrev) [inline fem/linearform.cpp]: Just noticed this: the trick with the `FakeTransformation` will not work with for `VectorFiniteElement`s (i.e. RT_* and ND_* elements) because there you need the real transformation to compute the basis functions in physical (not reference) space.

Comment (stefanozampini) [reply fem/linearform.cpp]: That's strange. I'm using it with Nedelec elements and it seems it's working.
Anyway, can you provide a fix for these cases?
